proposed/resource-limit.md

```diff
@@ -0,0 +1,843 @@
+
+- `Window`
+
+This represents the time span where the `ResourceLimit` is to be applied.
+
+1. `SlidingWindowRateLimiterOptions`
+
+- `ResourceLimit` - see `FixedWindowRateLimiterOptions`
+- `ResourceDepletedMode` - see `FixedWindowRateLimiterOptions`
+- `MaxQueueSize` - see `FixedWindowRateLimiterOptions`
+- `TickEventHandler` - see `FixedWindowRateLimiterOptions`
+- `Window` - see `FixedWindowRateLimiterOptions`
+- `SegmentsPerWindow`
+
+This represents how many segments each window is to be divided into.
+
+3. `TokenBucketRateLimiterOptions`
+
+- `ResourceLimit` - see `FixedWindowRateLimiterOptions`
+- `ResourceDepletedMode` - see `FixedWindowRateLimiterOptions`
+- `MaxQueueSize` - see `FixedWindowRateLimiterOptions`
+- `TickEventHandler` - see `FixedWindowRateLimiterOptions`
+- `ReplenishmentPeriod`
+
+This specifies how often replenishment occurs.
+
+- `TokensPerPeriod`
+
+This specifies how many tokens are restored each time replenishment occurs.
+
+4. `ConcurrencyLimiterOptions`
+
+- `ResourceLimit` - see `FixedWindowRateLimiterOptions`
+- `ResourceDepletedMode` - see `FixedWindowRateLimiterOptions`
+- `MaxQueueSize` - see `FixedWindowRateLimiterOptions`
+
+#### Channels
+
+Instead of modifying the existing `BoundedChannel` and `UnboundedChannel`, the proposal here is to create a new `ResourceLimitedChannel` along with a `ResourceLimitedChannelOptions`.
+
+```c#
+namespace System.Threading.Channels
+{
+    public sealed class ResourceLimitedChannelOptions : ChannelOptions
+    {
+        public ResourceLimiter WriteLimiter { get; set; }
+    }
+
+    internal sealed class ResourceLimitedChannel<T> : Channel<T> { }
+
+    public static class Channel
+    {
+        public static Channel<T> CreateResourceLimited<T>(LimitedChannelOptions options)
+    }
+}
+```
+
+The `ResourceLimitedChannel` represents a superset of the functionality of `BoundedChannel` allowing for configuration with both concurrency and rate limits. As a general overview, the `WriteLimiter` is queried on `TryWrite`, `WriteAsync` and `WaitToWriteAsync`. When a written item is enqueued to `_items`, a corresponding `Resource` is also added. Upon dequeue from `_items` from the `ChannelReader` or upon unblocking a reader, `Resource.Dispose` is invoked to release resources back to the write limiter. NOte that the `ResourceLimitedChannel` will only limit the producer side, i.e. writes. For details on proof of concept for the implementation, see [dotnet/runtime PoCs](#dotnetruntime-pocs).
+
+#### Pipelines
+
+TBD
+
+## Proof-of-concepts
+
+### Customized resource limiter based on ambient state that won't ship in box
+
+e.g. The limiter fails when CPU usage is above 90%
+
+```c#
+class CPUUsageLimiter : ResourceLimiter
+{
+    private PerformanceCounter _cpuUsageCounter = new PerformanceCounter("Processor", "% Processor Time", "_Total");
```

Comment:
This is windows specific

Reply:
Yea I'm illustrating a possibility here, it's certainly not how we would actually go about implementing this.